src/libraries/TwabLib.sol

```diff
@@ -18,435 +17,674 @@ import { ObservationLib, MAX_CARDINALITY } from "./ObservationLib.sol";
-   * @param _twabs          Individual user Observation recorded checkpoints passed as storage pointer
-   * @param _accountDetails User AccountDetails struct loaded in memory
-   * @param _startTime      Start of timestamp range as an epoch
-   * @param _endTime        End of timestamp range as an epoch
-   * @return uint256 Average balance of user held between epoch timestamps start and end
+   * @notice Looks up the oldest observation in the circular buffer.
+   * @param _observations The circular buffer of observations
+   * @param _accountDetails The account details to query with
+   * @return index The index of the oldest observation
+   * @return observation The oldest observation in the circular buffer
    */
-  function getAverageBalanceBetween(
-    ObservationLib.Observation[MAX_CARDINALITY] storage _twabs,
-    AccountDetails memory _accountDetails,
-    uint32 _startTime,
-    uint32 _endTime
-  ) internal view returns (uint256) {
-    uint32 _currentTime = uint32(block.timestamp);
-    _endTime = _endTime > _currentTime ? _currentTime : _endTime;
+  function getOldestObservation(
+    ObservationLib.Observation[MAX_CARDINALITY] storage _observations,
+    AccountDetails memory _accountDetails
+  ) internal view returns (uint16 index, ObservationLib.Observation memory observation) {
+    // If the TWAB is not initialized we go to the beginning of the TWAB circular buffer at index 0
+    if (_accountDetails.cardinality < MAX_CARDINALITY) {
+      index = 0;
+      observation = _observations[0];
+    } else {
+      index = _accountDetails.nextObservationIndex;
+      observation = _observations[index];
+    }
+  }
 
-    (uint16 oldestTwabIndex, ObservationLib.Observation memory oldTwab) = oldestTwab(
-      _twabs,
-      _accountDetails
+  /**
+   * @notice Looks up the newest observation in the circular buffer.
+   * @param _observations The circular buffer of observations
+   * @param _accountDetails The account details to query with
+   * @return index The index of the newest observation
+   * @return observation The newest observation in the circular buffer
+   */
+  function getNewestObservation(
+    ObservationLib.Observation[MAX_CARDINALITY] storage _observations,
+    AccountDetails memory _accountDetails
+  ) internal view returns (uint16 index, ObservationLib.Observation memory observation) {
+    index = uint16(
+      RingBufferLib.newestIndex(_accountDetails.nextObservationIndex, MAX_CARDINALITY)
     );
-    (uint16 newestTwabIndex, ObservationLib.Observation memory newTwab) = newestTwab(
-      _twabs,
-      _accountDetails
+    observation = _observations[index];
+  }
+
+  /**
+   * @notice Looks up a users balance at a specific time in the past.
+   * @dev If the time is not an exact match of an observation, the balance is extrapolated using the previous observation.
+   * @dev Ensure timestamps are safe using isTimeSafe.
```

Comment:
So before calling this function, we need to call `isTimeSafe`?
It doesn't look like `isTimeSafe` is used anywhere in the codebase and since it's an internal function, we can't call it from another contract without inheriting this lib.

Reply:
It's more for external users who want to guarantee the time is safe. Since we'll be querying at specific times lined up with periods, we know it will be and don't need to waste the gas.

I forgot to expose it in the TWAB Controller.